### DISCUSSION TITLE
Running a process in a singularity container without bash?

### DISCUSSION THREAD

**Discussion by davidmasp** [nextflow-io/nextflow#5410] [Q&A] [answered]
Hi,

I am trying to run a process in a container that does not have bash (based on alpine). This one to be precise:

https://github.com/typst/typst/blob/main/Dockerfile

When "shelling" into the image I can interact with it using the `/bin/sh` shell.

I have tried to add `process.shell = ['/bin/sh']` but still my process fails because the call to bash is tied to the `singularity exec`. 

```
[blah blah] singularity exec [blah blah] container.img **/bin/bash** -c "[blah blah] /bin/sh"
```

I have seen this issue [here](https://github.com/nextflow-io/nextflow/issues/974) so I am assuming there's no easy fix. My question is, what do people do in this scenario? Do they just recreate the container and install bash somehow? Is there a way to change the singularity shell even if not using all nextflow features? I don't care about execution reports atm, just need singularity and slurm.

Thanks in advance.

**Comment by pditommaso** [comment]
A nextflow job \*is* a Bash script. therefore I see this hard. 

This problem is solved by using Conda package + Wave = on-demand containers 

https://www.nextflow.io/docs/latest/wave.html#build-conda-based-containers

**Reply by davidmasp** [reply]
mmmh I understand that. I don't have a problem on bash running the job (job as in the `.command.run` script?) but in the `singularity exec` line within the job. That being said, I ofc don't know enough of nextflow internals to fully understand the problem.

Many "official" containers that I am working on seem to rely on alpine which doesn't have bash.

i am trying to avoid conda at all cost so I am not sure I am ready for switching to this solution. I also probably don't know enough about wave, isn't it a fully cloud thing? I would like to run it locally/hpc and avoid cloud.

**Reply by pditommaso** [reply, marked as answer]
Along the tool, nextflow uses Bash (and `ps`) also to extract runtime metrics, e.g. mem & cpu usage, etc. 


> I also probably don't know enough about wave, isn't it a fully cloud thing? I would like to run it locally/hpc and avoid cloud.

It's a cloud thing, but you can build the container, wither with nexflow or the `wave` cli tool, and download it locally (see also the `nextflow inspect`) command at thsi regard 

```
» wave --conda bwa=0.7.18 --singularity -F 
oras://community.wave.seqera.io/library/bwa:0.7.18--1a0fbdc844bce6b9
```

**Reply by davidmasp** [reply]
Cool! Thanks for the input!
